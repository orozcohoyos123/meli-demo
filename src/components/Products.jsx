import React from "react";
import PropTypes from "prop-types";
import shipLogo from '../assets/ic_shipping.png';
import { Link } from "react-router-dom";
import "./styles/Products.scss";
import "./styles/ProductsSkeleton.scss"

const Products = ({ products, loading }) => {
  return (
    <section className="detail-wrapper col-11">
      {loading ? (
        <ol className="detail-list">
          {[1, 1, 1, 1].map((number, index) => (
            <li key={index}>
              <div className="items-wrapper">
                <div className="items-image-wrapper">
                  <div className="items-skeleton__img items__image"></div>
                </div>
                <div className="items-content-wrapper">
                  <div className="items-information row">
                    <div className="col">
                      <div className="items-information-price items-information-price-skeleton"></div>
                    </div>
                    <span className="items-information-location items-information-location-skeleton col-sm-auto"></span>
                    <div className="col col-sm-1"></div>
                  </div>
                  <div className="items-description items-description-skeleton"></div>
                  <div className="items-description items-description-skeleton"></div>
                </div>
              </div>
            </li>
          ))}
        </ol>
      ) : (
        <ol className="detail-list">
          {products && products.map((product, index) => {
            return (
              <li key={`${product.id}_${index}`} >
                <div className="items-wrapper">
                  <div className="items-image-wrapper">
                    <Link to={`/items/${product.id}`}>
                      <img className="items__image" src={product.picture} alt={product.id} />
                    </Link>
                  </div>
                  <div className="items-content-wrapper">
                    <div className="items-information row">
                      <Link to={`/items/${product.id}`} className="items-information-price col">
                        {product.price && `${product.price.currency} ${new Intl.NumberFormat("de-DE").format(product.price.amount)}`}
                        {product.free_shipping && (
                          <img className="items-ship-logo" src={shipLogo} alt="" />
                        )}
                      </Link>
                      <span className="items-information-location col-sm-auto">
                        {product.location}
                      </span>
                      <div className="col col-sm-1"></div>
                    </div>
                    <div className="items-description">{product.title}</div>
                  </div>
                </div>
              </li>
            );
          })}
        </ol>
      )
      }
    </section>
  );
};

Products.propTypes = {
  products: PropTypes.array,
  loading: PropTypes.bool
};

export default Products;
