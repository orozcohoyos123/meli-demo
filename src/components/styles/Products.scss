@import '../../styles/CommonColors.scss';

.detail-wrapper {
    background: white;
    border-radius: 4px;
    padding: 0;
}

.detail-list {
    list-style: none;
    margin: 0;
    padding: 0;
}

.detail-list li {
    display: flex;
    flex-direction: column;
    list-style: none;
    padding: 16px;
}

.detail-list li:last-child {
    padding-bottom: 0 ;
}

.items-wrapper {
   display: flex;
   border-bottom: 1px solid $light-gray;
}

.items-image-wrapper {
    box-sizing: content-box;
    padding-right: 16px;
}

.items__image {
    border-radius: 4px;
    height: 180px;
    object-fit: contain;
    width: 180px;
}

.items-content-wrapper {
    display: flex;
    flex-direction: column;
    width: calc(100% - 180px);
}

.items-information {
    padding-top: 16px;
    align-items: baseline;
}

.items-information-price {
    font-size: 24px;
    color: #333333;
}

.items-information-location {
    font-size: 12px;
    color: $middle-gray;
}

.items-description {
    padding-top: 16px;
}