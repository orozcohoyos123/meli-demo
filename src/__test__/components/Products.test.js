import React from 'react';
import { shallow, mount } from 'enzyme';
import { create } from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom';
import Products from '../../components/Products';

const mockProducts = [
    {
      id: "MLA920543431",
      condition: "new",
      free_shipping: true,
      location: "Monte Chingolo",
      picture: "http://http2.mlstatic.com/D_810459-MLA44182891230_112020-I.jpg",
      price:{
        currency: "ARS",
        decimals: 0,
        amount: 74999
      },
      title: "Sony Playstation 4 Slim 1tb Marvel's Spider-man/horizon Zero Dawn Complete Edition/ratchet & Clank  Color Negro"
    }
]

describe('<Products />', () => {
    test('Debe renderizar los skeleton items del componente Products sin mock de productos y con parametro cargando en true.', () => {
        const products = shallow(<Products loading={true}/>);
        const skeleton_items = products.find('.items-skeleton__img');
        const items = products.find('img');

        expect(products.length).toEqual(1);
        expect(skeleton_items.length).toEqual(4);
        expect(items.length).toEqual(0);
    }) 

    test('No debe renderizar items del componente Products sin mock de productos y con parametro cargando en false.', () => {
        const products = shallow(<Products loading={false}/>);
        const items = products.find('items-wrapper');

        expect(products.length).toEqual(1);
        expect(items.length).toEqual(0);
    }) 

    test('Debe renderizar items del componente Products con mock de productos y con parametro cargando en false.', () => {
        const products = mount(
            <MemoryRouter>
                <Products products={mockProducts} loading={false}/>
            </MemoryRouter>
        );
        const skeleton_items = products.find('.items-skeleton__img');
        const items = products.find('img').filter('.items__image');

        expect(products.length).toEqual(1);
        expect(skeleton_items.length).toEqual(0);
        expect(items.length).toEqual(mockProducts.length);
    }) 

    test('No debe renderizar items del componente Products con mock de productos y con parametro cargando en true.', () => {
        const products = shallow(<Products loading={true}/>);
        const skeleton_items = products.find('.items-skeleton__img');
        const items = products.find('img');

        expect(products.length).toEqual(1);
        expect(skeleton_items.length).toEqual(4);
        expect(items.length).toEqual(0);
    })  
});

describe('Products sanpshot', () => { 
    test('Comprobar UI componente Products', () => {
        const products = create(
            <MemoryRouter>
                <Products products={mockProducts} loading={false}/>
            </MemoryRouter>
        );
        
        expect(products.toJSON()).toMatchSnapshot();
    })
});
