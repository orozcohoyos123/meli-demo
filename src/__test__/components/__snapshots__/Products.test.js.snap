// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products sanpshot Comprobar UI componente Products 1`] = `
<section
  className="detail-wrapper col-11"
>
  <ol
    className="detail-list"
  >
    <li>
      <div
        className="items-wrapper"
      >
        <div
          className="items-image-wrapper"
        >
          <a
            href="/items/MLA920543431"
            onClick={[Function]}
          >
            <img
              alt="MLA920543431"
              className="items__image"
              src="http://http2.mlstatic.com/D_810459-MLA44182891230_112020-I.jpg"
            />
          </a>
        </div>
        <div
          className="items-content-wrapper"
        >
          <div
            className="items-information row"
          >
            <a
              className="items-information-price col"
              href="/items/MLA920543431"
              onClick={[Function]}
            >
              ARS 74.999
              <img
                alt=""
                className="items-ship-logo"
                src="ic_shipping.png"
              />
            </a>
            <span
              className="items-information-location col-sm-auto"
            >
              Monte Chingolo
            </span>
            <div
              className="col col-sm-1"
            />
          </div>
          <div
            className="items-description"
          >
            Sony Playstation 4 Slim 1tb Marvel's Spider-man/horizon Zero Dawn Complete Edition/ratchet & Clank  Color Negro
          </div>
        </div>
      </div>
    </li>
  </ol>
</section>
`;
